Xamarin.Forms.Core/IndicatorView.cs

```diff
@@ -0,0 +1,256 @@
+			set => SetValue(IndicatorColorProperty, value);
+		}
+
+		public Color SelectedIndicatorColor
+		{
+			get => (Color)GetValue(SelectedIndicatorColorProperty);
+			set => SetValue(SelectedIndicatorColorProperty, value);
+		}
+
+		public double IndicatorSize
+		{
+			get => (double)GetValue(IndicatorSizeProperty);
+			set => SetValue(IndicatorSizeProperty, value);
+		}
+
+		public IEnumerable ItemsSource
+		{
+			get => (IEnumerable)GetValue(ItemsSourceProperty);
+			set => SetValue(ItemsSourceProperty, value);
+		}
+
+		void ResetIndicatorStyles()
+		{
+			try
+			{
+				BatchBegin();
+				ResetIndicatorStylesNonBatch();
+			}
+			finally
+			{
+				BatchCommit();
+			}
+		}
+
+		void ResetIndicatorCount(int oldCount)
+		{
+			try
+			{
+				BatchBegin();
+				if (oldCount < 0)
+				{
+					oldCount = 0;
+				}
+
+				if (oldCount > Count)
+				{
+					RemoveRedundantIndicatorItems();
+					return;
+				}
+
+				AddExtraIndicatorItems();
+			}
+			finally
+			{
+				ResetIndicatorStylesNonBatch();
+				BatchCommit();
+			}
+		}
+
+		void ResetIndicatorStylesNonBatch()
+		{
+			foreach (var child in Items)
+			{
+				ApplyColor(child as View);
+			}
+
+			IndicatorLayout.IsVisible = Count > 1 || !HidesForSingleIndicator;
+		}
+
+		void ResetItemsSource(IEnumerable oldItemsSource)
+		{
+			if (oldItemsSource is INotifyCollectionChanged oldObservableCollection)
+			{
+				oldObservableCollection.CollectionChanged -= OnCollectionChanged;
+			}
+
+			if (ItemsSource is INotifyCollectionChanged observableCollection)
+			{
+				observableCollection.CollectionChanged += OnCollectionChanged;
+			}
+
+			OnCollectionChanged(ItemsSource, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+		}
+
+		void AddExtraIndicatorItems()
+		{
+			var oldCount = Items.Count;
+			for (var i = 0; i < Count - oldCount && i < MaximumVisibleCount - oldCount; ++i)
+			{
+				var size = IndicatorSize > 0 ? IndicatorSize : 10;
+				var indicator = IndicatorTemplate?.CreateContent() as View ?? new Frame
+				{
+					Padding = 0,
+					HasShadow = false,
+					BorderColor = Color.Transparent,
+					VerticalOptions = LayoutOptions.Center,
+					HorizontalOptions = LayoutOptions.Center,
+					WidthRequest = size,
+					HeightRequest = size,
+					CornerRadius = (float)size / 2
+				};
+				var itemTapGesture = new TapGestureRecognizer();
+				itemTapGesture.Tapped += (tapSender, tapArgs) => Position = Items.IndexOf(tapSender as View);
+				indicator.GestureRecognizers.Add(itemTapGesture);
+				Items.Add(indicator);
+			}
+		}
+
+		void RemoveRedundantIndicatorItems()
+		{
+			foreach (var item in Items.Where((v, i) => i >= Count).ToArray())
+			{
+				Items.Remove(item);
+			}
+		}
+
+		void ApplyColor(View view)
+		{
+			var index = Items.IndexOf(view);
+			if (index == Position)
+			{
+				ApplySelectedColor(view);
+				return;
+			}
+			ApplyRegularColor(view);
+		}
+
+		void ApplySelectedColor(View indicatorItemView)
```

Comment:
Do we really need `ApplySelectedColor`, `ApplyRegularColor`, and `GetColorOrDefault`? These are one-line methods and we can use their code directly in the callers instead of extracting everything into separate methods.

Reply:
@adrianknight89 you are right. They are presented because previously they were protected virtual. I decided to hide them for now. Of course, I can drop them)

Reply:
My opinion is dropping them would be better.

Reply:
done